include/ocpp/common/message_queue.hpp

```diff
@@ -340,46 +381,49 @@ template <typename M> class MessageQueue {
+
+                if (selected_normal_message_it != normal_message_queue.end()) {
+                    message = *selected_normal_message_it;
+                    queue_type = QueueType::Normal;
                 }
 
-                if (!this->transaction_message_queue.empty()) {
-                    auto& transaction_message = this->transaction_message_queue.front();
-                    EVLOG_debug << "transaction msg timestamp: " << transaction_message->timestamp;
-                    if (message == nullptr) {
-                        if (transaction_message->timestamp <= now) {
-                            EVLOG_debug << "transaction message timestamp <= now";
-                            message = transaction_message;
-                            queue_type = QueueType::Transaction;
-                        }
-                    } else {
-                        if (transaction_message->timestamp <= message->timestamp and
-                            message->messageType != M::BootNotification) {
-                            EVLOG_debug << "transaction message timestamp <= normal message timestamp";
-                            message = transaction_message;
-                            queue_type = QueueType::Transaction;
-                        } else {
-                            EVLOG_debug << "Prioritizing newer normal message over older transaction message";
-                        }
-                    }
+                // Find the first allowed transaction message
+                auto selected_transaction_message_it =
+                    std::find_if(transaction_message_queue.begin(), transaction_message_queue.end(),
+                                 [&](const std::shared_ptr<ControlMessage<M>>& msg) {
+                                     if (!this->allowed_to_send_message(msg)) {
+                                         return false;
+                                     }
+                                     // no message selected from normal message queue, so select transaction message
+                                     if (message == nullptr) {
+                                         return true;
+                                     }
+                                     // message from normal message queue is BootNotification, this is prioritized
+                                     if (message->messageType == M::BootNotification) {
+                                         return false;
+                                     }
+                                     // transaction messages is older than normal message, so select transaction message
+                                     if (msg->timestamp <= message->timestamp) {
+                                         return true;
+                                     }
+                                     return false;
+                                 });
```

Comment:
Due to the high indentation level, the formatting is a bit ugly.  What you can do in this case, and what in general is a good idea for *longer* lambdas, define them beforehand and give them a *self documenting* name like:
```
auto is_transaction_message_available = [&](const std::shared_ptr<ControlMessage<M>>& msg) {
  ...
};
const auto transaction_message_it = std::find_if(std::begin(transaction_message_queue), std::end(transaction_message_queue), is_transaction_message_available);
```
The name `is_transaction_message_available` is of course not completely correct, because it prioritizes some of the *normal* messages.  Also using `msg` and `message` is a bit unfortunate.  Also note, that starting with c++14 you can use aliases in the capture list, so you could have written
`auto ... = [this, &normal_message = message](const ... transaction_message) {}`.